<a href="https://colab.research.google.com/github/kankeinai/physics_classes/blob/main/Assignment_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PHYS270 Assigmment 1: Lists and Numpy Arrays**
Student Name: Begantsova Milana



---


## Abstract




---


##Introduction



---


##Method



---


##Experiment
At first we should import all neccessary libraries needed while the experiment.

In [ ]:
import numpy as np
from random import random
from math import sqrt

The experiment consists of three parts. Shortly, in the Part (a) we will generate two lists with different 1000 random float numbers. These two lists will be generated using function random( ). Then in Part (b), we should measure time needed to generate a different list elements of which will be calculated according to defined formula using elements from list a and b. In the last Part (c), we will do the same but instead of lists, numpy arrays will be used. 

###Part (a)
In the first part, we should randomly create two python lists. Lets call them a and b. They should be of the $size=1000$.

In [ ]:
size = 1000

All of the elements will be generated with function random(  ). To generate those two lists, I will use python shorthand.

In [ ]:
list_a = [random() for i in range(size)]
list_b = [random() for i in range(size)]

###Part (b)
In this part, we should generate **list c**. The elements of this list should be calculated according to the next formula:
$$c[i]=\sqrt{a[i]^2+b[i]^2}$$
So, we should elementwise square elements of **a** and **b**, then sum then one by one, and end up calculating their square root.

As it has already been discussed in the method section, I will use jupiter notebook built-in function **%%timeit** to measure time needed to generate **list c**. 

To generate list c, I will again use python shorthand for creating lists, but this time we work with elements of both lists. Python bult-in lists does not support adding two lists elementwise, so I will use function **zip( )** to create list of tuples (pairs) of elements **a** and **b**. Then in the same manner as in Part (a), I use needed formula to calculate elements of **list c**.

In [ ]:
%%timeit 
list_c = [sqrt(item_a**2 + item_b**2) for item_a, item_b in zip(list_a, list_b)]

1000 loops, best of 3: 216 µs per loop


As you can see, it took $216 \mu s$ to generate list of 1000 elements according to formula (1). It is not bad. But 
in data science, experts work with billions of recording with many features (in this experiment we have only two features: element from list **a** and element from list **b**), so this is quite frustrating result. We generate this list using for-loop, so the time needed to acomplish the task will be linearly proportional to number of elements in a list. In addition, (про то как хранятся данные в питоне).

###Part (c)
In this section we will do the same task as in Part (a) and (b), but instead of using lists, we will use numpy arrays.

In data science, we usually work with numbers. So there is no need in such generic data structers as lists. They not only have long time of accessing elements, but also don't have efficient implementation of all functions that we need, such as element-wise additon, power, substraction and etc. 

So, there is an alternative: numpy arrays. 

They are created in a such way to be able to make all kind of operations that we can do with matrices. Numpy arrays are based on C-style arrays, which are more low level and fast.

Our goal, in this experiment, is to conclude whatever they are more efficient then python lists.

I will not generate elements for the second time, as numpy library has built-in function of translating lists to arrays. If we compare how two data structures work with **the same elements**, the accuracy of the experiment should be highier.

In [ ]:
array_a = np.array(list_a)
array_b = np.array(list_b)

As I have already said, when we work with numpy arrays there are ready-to-use functions for finding power, addition and square root extraction. So, there is no need to use a for loop to access elements one by one. We work with matrices instead. To function work properly, arrays should be of the same size.

In [ ]:
%%timeit
array_c = np.sqrt(np.add(np.power(array_a, 2),np.power(array_b, 2)))

The slowest run took 19.11 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 25.5 µs per loop


As you can see, it took $25.5 \mu s$ to generate an **array c**, which is almost 9 times faster then with lists.

To end this experiment, let me construct a summative table to look at the results and make sure that elements of **list c** and **array c** are the same. To construct a summative table, I will use pandas library.

In [ ]:
import pandas as pd

df=pd.DataFrame({'list a':list_a, 'list b': list_b, 'list c': list_c, 'array c':array_c})
df

list a    list b    list c   array c
0    0.715413  0.445999  0.807233  0.807233
1    0.280579  0.280654  0.790237  0.790237
2    0.943882  0.371968  0.991286  0.991286
3    0.564539  0.726863  1.174707  1.174707
4    0.335098  0.046797  1.272573  1.272573
..        ...       ...       ...       ...
995  0.488877  0.152869  1.025443  1.025443
996  0.803706  0.966243  1.223005  1.223005
997  0.380691  0.440158  0.203872  0.203872
998  0.677767  0.596759  0.657871  0.657871
999  0.704230  0.946341  0.478978  0.478978

[1000 rows x 4 columns]

It seems that everything works corectly for both lists and numpy arrays.



---


##Conclusion